### PR TITLE
[RavenDB-12658] refactoring + tests

### PR DESCRIPTION
**Pull request by Or-Geva**

### CODE REVIEW

**Review by grisha-kotler — APPROVED** [review]

File: src/Raven.Client/Documents/Session/DocumentSessionAttachmentsBase.cs [hunk 1]
```diff
@@ -71,20 +71,30 @@ public void Store(string documentId, string name, Stream stream, string contentT
...
             Store(document.Id, name, stream, contentType);
         }
 
-        protected void ThrowEntityNotInSession(object entity)
+        protected void ThrowEntityNotInSessionOrMissingId(object entity)
         {
             throw new ArgumentException($"{entity} is not associated with the session. Use documentId instead or track the entity in the session.", nameof(entity));
         }
 
+        protected void ThrowEntityNotInSession(object entity)
```
Comment (ppekrol) [inline src/Raven.Client/Documents/Session/DocumentSessionAttachmentsBase.cs:87]: I do not like it, create 2 overloads

**Review by ppekrol — CHANGES_REQUESTED** [review]

**Comment by Or-Geva** [comment]
Better error message when trying to get names of attachment from a document on in the session

File: src/Raven.Client/Documents/Session/DocumentSessionAttachmentsBase.cs (hunk 1 above)
Comment (Or-Geva) [reply src/Raven.Client/Documents/Session/DocumentSessionAttachmentsBase.cs:87]: I cannot overload with an object,
pulled out the if to implementation of getNames

File: src/Raven.Client/Documents/Session/DocumentSessionAttachmentsBase.cs [hunk 2]
```diff
@@ -27,6 +27,9 @@ public AttachmentName[] GetNames(object entity)
             if (entity == null)
                 throw new ArgumentNullException(nameof(entity));
 
+            if (entity is string)
+                throw new ArgumentException($"You need to track the entity in the session for document id {entity}.", nameof(entity));
```
Comment (ppekrol) [inline src/Raven.Client/Documents/Session/DocumentSessionAttachmentsBase.cs]: Can we be more explicit here? I've read that few times and I'm still having issues to understand what would be the action that I should take to not get this exception when this method is used.

Comment (Or-Geva) [reply src/Raven.Client/Documents/Session/DocumentSessionAttachmentsBase.cs]: GetNames is the only one that has no overload by 'documentId', as opposed to other attachments' methods such as get, store, etc.

Comment (ppekrol) [reply src/Raven.Client/Documents/Session/DocumentSessionAttachmentsBase.cs]: I understand, but the exception message is not 100% clear

Comment (Or-Geva) [reply src/Raven.Client/Documents/Session/DocumentSessionAttachmentsBase.cs]: Done.

**Review by ppekrol — APPROVED** [review]

**Pull request merged**